### ISSUE TITLE
Add API Progress Callbacks for Sizing Calculations

### ISSUE THREAD

**Issue by TiejunWu** [NREL/EnergyPlus#10688]
Issue overview
--------------
Currently for 3rd party applications that uses the EnergyPlus API, the API provides a progress callback after each simulation day [here](https://github.com/NREL/EnergyPlus/blob/07dbb5b230ad78298ea3d9a4a3306317fde6bef5/src/EnergyPlus/DisplayRoutines.cc#L135). However, in an EnergyPlus simulation, there are other steps prior to the primary simulation that could take significant time to complete, but there is no progress callbacks for those steps, this makes it hard for these 3rd party applications to know the progress.

In the attached [Example EnergyPlus Input Data File](https://github.com/TiejunWu/EnergyPlus-IDF-Files/blob/master/ProgressReportExample.idf), the SimulationControl Object specifies the program to first do the sizing calculation and then run simulation for sizing periods. This input file has 13 DesignDays. 

![image](https://github.com/user-attachments/assets/c51b5449-f246-49c4-aca4-8d4c69138f93)

In the calculation, after the initialization steps, following are the sequence of events during the calculation:

1. EnergyPlus will perform Zone Sizing Simulation for each designday
2. EnergyPlus will calculate System Sizing for each designday
3. EnergyPlus will run simulation for each designday (Primary Simulation). 

Currently the API Progress Callback is provided only for step#3 after simulation for each designday is finished. For step#1, the zone sizing simulation for each designday could take about the same time duration as the primary simulation. Without any progress report callbacks, it would make the simulation look like stuck. 

Another issue is that the callback reported progress percentage number seems to be based on the assumption that there are always 365 simulation days. So for the attached example the progress callback reported percentage are:

          0
          1
          1
          1
          1
          2
          2
          2
          2
          3
          3
          3
          3
          100

In summary, we would like to see the following improvements:
1.  In addition to the existing progress callback during the primary simulation.  Add progress reporting callback after each designday zone sizing simulation (the system sizing calculation and plant sizing calculation seems to be much faster, not sure if callbacks are necessary).
2.  Revise the algorithm that calculates the progress percentage number based on the actual number of simulation days rather than just use 365 as the denominator all the time.

### Details
Some additional details for this issue (if relevant):
 - Platform (Windows 10, 64 bit)
 - Version of EnergyPlus (EnergyPlus 23.2)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to EnergyPlus Defect Complexity (Github Project)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

**Comment by jmarrec** [comment]
> Revise the algorithm that calculates the progress percentage number based on the actual number of simulation days rather than just use 365 as the denominator all the time.

I don't think that is true? https://github.com/NREL/EnergyPlus/blob/acda84e38db4e8aecbc178cb970055d3721ee321/src/EnergyPlus/DisplayRoutines.cc#L135-L164

**Comment by jmarrec** [comment]
For reporting during sizing, I'm not sure I follow what you're after.

You have a variety of callbacks already.

callback_begin_new_environment -> new environment, that includes both design days and run period sims
callback_after_new_environment_warmup_complete -> warmup is done
callback_end_zone_sizing -> zone sizing is done
callback_end_system_sizing -> system sizing is done

Just figure out the way you want your progress bar to move. `progress_callback` should be 1 to 100, so maybe instead do something that goes 1-130 total:

callback_begin_new_environment -> 1
callback_after_new_environment_warmup_complete -> 10
callback_end_zone_sizing -> 20
callback_end_system_sizing -> 30
progress_callback -> i + 30 (31 to 130)

You can figure out how many design days you have via `len(data_transfer.get_object_names("SizingPeriod:DesignDay"))`, and figure a global marching of progress accross the entire sim.



Given the above, do you still have a request?

**Comment by DavidHollman** [comment]
Here are some example IDF files for which `progress_callback` produces output like what @TiejunWu mentioned:

- [LoadCalculation.zip](https://github.com/user-attachments/files/18215415/LoadCalculation.zip) and [SystemSizing.zip](https://github.com/user-attachments/files/18215416/SystemSizing.zip) both increment only a few times and then jump to 100. Seems like `CurrentSimDay` in the code sample your pointed out must not ever reach the value of `TotalSimDays`.

- [WholeYearSimulation.zip](https://github.com/user-attachments/files/18215417/WholeYearSimulation.zip) counts up to nearly 100; output values:
  - `0
1
1
1
1
2
2
2
2
3
3
3
3
4
4
4
4
5
5
5
6
6
6
6
7
7
7
7
8
8
8
8
9
9
9
10
10
10
10
11
11
11
11
12
12
12
12
13
13
13
13
14
14
14
15
15
15
15
16
16
16
16
17
17
17
17
18
18
18
19
19
19
19
20
20
20
20
21
21
21
21
22
22
22
22
23
23
23
24
24
24
24
25
25
25
25
26
26
26
26
27
27
27
28
28
28
28
29
29
29
29
30
30
30
30
31
31
31
31
32
32
32
33
33
33
33
34
34
34
34
35
35
35
35
36
36
36
37
37
37
37
38
38
38
38
39
39
39
39
40
40
40
40
41
41
41
42
42
42
42
43
43
43
43
44
44
44
44
45
45
45
46
46
46
46
47
47
47
47
48
48
48
48
49
49
49
49
50
50
50
51
51
51
51
52
52
52
52
53
53
53
53
54
54
54
54
55
55
55
56
56
56
56
57
57
57
57
58
58
58
58
59
59
59
60
60
60
60
61
61
61
61
62
62
62
62
63
63
63
63
64
64
64
65
65
65
65
66
66
66
66
67
67
67
67
68
68
68
69
69
69
69
70
70
70
70
71
71
71
71
72
72
72
72
73
73
73
74
74
74
74
75
75
75
75
76
76
76
76
77
77
77
78
78
78
78
79
79
79
79
80
80
80
80
81
81
81
81
82
82
82
83
83
83
83
84
84
84
84
85
85
85
85
86
86
86
87
87
87
87
88
88
88
88
89
89
89
89
90
90
90
90
91
91
91
92
92
92
92
93
93
93
93
94
94
94
94
95
95
95
96
96
96
96
97
100` (366 total)

- All of these examples report the _same value_ multiple times. That may be because its truncating to integer, but regardless it means that the actual value isn't very useful. 

Based on that, recommendations for `progress_callback`:
 - Always return increasing values (not necessarily integers)
 - Smoothly count up to 100%

---

As for the other callbacks you mentioned - thanks, those look useful and we'll try them out.  

That said, they may not be sufficiently granular to help much in cases of longer runtimes. That's what the request to add more progress reporting throughout the entire run was based on, to achieve a more even granularity of progress tracking overall. As far as it goes, I think this is the more important issue.
